### PR TITLE
adding InterpolatedIsotonicRegression

### PR DESCRIPTION
**Pull request by cranmer**
not sure if this is the right place to put this class, or how to modify CalibratedClassifierCV to use it.

### CODE REVIEW

**Comment by cranmer** [comment]
Some updates to the InterpolatedIsotonicRegression class and a failed attempt to integrate it into the calibration

File: carl/learning/calibration.py [hunk 1]
```diff
@@ -165,3 +181,90 @@ def clone(self):
...
+        into piecewise linear model
+    """
+
+    def __init__(self, y_min=None, y_max=None, increasing=True,
+                 out_of_bounds='nan'):
+        self.y_min = y_min
+        self.y_max = y_max
+        self.increasing = increasing
+        self.out_of_bounds = out_of_bounds
+        self.iso_ = IsotonicRegression(y_min=y_min, y_max=y_max,
```
Comment (glouppe) [inline carl/learning/calibration.py]: By scikit-learn conventions, can you move this into `fit`?

File: carl/learning/calibration.py [hunk 2]
```diff
@@ -165,3 +181,90 @@ def clone(self):
...
+        change_mask1[-1] = True
+        change_mask2[0] = True
+        change_mask2[-1] = True
+
+        self.iso_interp1_ = interp1d(X[change_mask1],p[change_mask1])
+        self.iso_interp2_ = interp1d(X[change_mask2],p[change_mask2])
+
+        self.xmin_ = np.max((np.min(X[change_mask1]),np.min(X[change_mask2])))
+        self.xmax_ = np.min((np.max(X[change_mask1]),np.max(X[change_mask2])))
+
```
Comment (glouppe) [inline carl/learning/calibration.py]: `fit` should return `self`

**Comment by glouppe** [comment]
Could you add this new method to the tests? It should be limited to adding "interpolated-isotonic" to https://github.com/diana-hep/carl/blob/master/tests/learning/test_calibration.py#L78

File: carl/learning/calibration.py [hunk 3]
```diff
@@ -53,12 +56,21 @@ def _fit_calibrators(self, df0, df1):
                                     range=[(df_min, df_max)],
                                     interpolation="linear")
 
+        elif self.method == "interpolated-isotonic":
+            self.interp_iso_ = InterpolatedIsotonicRegression()
+
+            Xtemp = np.hstack((df0.reshape(-1,),df1.reshape(-1,)))
```
Comment (glouppe) [inline carl/learning/calibration.py]: You can use `ravel()` instead of `reshape(-1,)`, it is easier to read.

File: carl/learning/calibration.py [hunk 4]
```diff
@@ -53,12 +56,21 @@ def _fit_calibrators(self, df0, df1):
...
+            self.interp_iso_.fit(Xtemp.reshape(-1,),
+                np.hstack((np.zeros(df0.size),1.+np.zeros(df1.size))))
```
Comment (glouppe) [inline carl/learning/calibration.py]: 1.+np.zeros(size) => np.ones(size)

File: carl/learning/calibration.py [hunk 5]
```diff
@@ -165,3 +181,90 @@ def clone(self):
...
+        change_mask2[-1] = True
+
+        self.iso_interp1_ = interp1d(X[change_mask1],p[change_mask1])
+        self.iso_interp2_ = interp1d(X[change_mask2],p[change_mask2])
+
+        self.xmin_ = np.max((np.min(X[change_mask1]),np.min(X[change_mask2])))
+        self.xmax_ = np.min((np.max(X[change_mask1]),np.max(X[change_mask2])))
+
+
+    def transform(self, T):
```
Comment (glouppe) [inline carl/learning/calibration.py]: T -> X

File: carl/learning/calibration.py [hunk 6]
```diff
@@ -53,12 +56,21 @@ def _fit_calibrators(self, df0, df1):
                                     range=[(df_min, df_max)],
                                     interpolation="linear")
 
+        elif self.method == "interpolated-isotonic":
+            self.interp_iso_ = InterpolatedIsotonicRegression()
+
+            Xtemp = np.hstack((df0.reshape(-1,),df1.reshape(-1,)))
+            self.interp_iso_.fit(Xtemp.reshape(-1,),
+                np.hstack((np.zeros(df0.size),1.+np.zeros(df1.size))))
+            calibrator0, calibrator1 = clone(self.method), clone(self.method)
```
Comment (glouppe) [inline carl/learning/calibration.py]: No need for cloning. Just directly `return None, None` after `self.interp_iso_.fit`

**Comment by glouppe** [comment]
Not critical, but to enforce good coding practices throughout the codebase, could you check PEP8 and PEP257 issues? (you can run `flake8` to check these)

File: carl/learning/calibration.py [hunk 7]
```diff
@@ -143,6 +156,9 @@ def predict_proba(self, X):
         if self.method in ("isotonic", "sigmoid"):
             return self.classifiers_[0].predict_proba(X)
 
+        elif self.method == "interpolated-isotonic":
+            self.interp_iso_.predict( self.classifiers_[0].predict_proba(X)[:,1] )
```
Comment (glouppe) [inline carl/learning/calibration.py]: You missed the `return` keyword here, hence the `NoneType` error.

**Comment by cranmer** [comment]
ok, basically working now. Here are some results:
https://github.com/cranmer/carl-notebooks/blob/iso-interp/Approximating%20likelihood%20ratios%20with%20calibrated%20classifiers.ipynb

**Comment by glouppe** [comment]
Merging this, thanks for the added calibration method!

**Pull request merged**